### PR TITLE
Add fuzz tests for in place array BPT dropping

### PR DESCRIPTION
**Pull request by TomAFrench**
This PR adds fuzz tests for the inplace array shortening trick with assembly. We're likely going to be inlining this assembly block where necessary so I haven't broken it out into a library for now.

### CODE REVIEW

File: pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        // Note that this requires tokens.length > 0 otherwise the array length will underflow.
+        assembly {
+            mstore(add(tokens, 32), sub(mload(tokens), 1))
+            tokens := add(tokens, 32)
+        }
+
+        emit log_named_array("Actual tokens without BPT", tokens);
+
+        for (uint256 i = 0; i < expectedTokensWithoutBpt.length; i++) {
```
Comment (nventuro) [inline pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol:30]:
```suggestion
        assertEq(tokens.length, expectedTokensWithoutBpt.length);
        for (uint256 i = 0; i < expectedTokensWithoutBpt.length; i++) {
```

File: pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+        // Note that this requires balances.length > 0 otherwise the array length will underflow.
+        assembly {
+            mstore(add(balances, 32), sub(mload(balances), 1))
+            balances := add(balances, 32)
+        }
+
+        emit log_named_array("Actual balances without BPT", balances);
+
+        for (uint256 i = 0; i < expectedBalancesWithoutBpt.length; i++) {
```
Comment (nventuro) [inline pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol:54]:
```suggestion
        assertEq(tokens.length, expectedTokensWithoutBpt.length);
        for (uint256 i = 0; i < expectedBalancesWithoutBpt.length; i++) {
```

File: pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        emit log_named_array("Actual tokens without BPT", tokens);
+
+        for (uint256 i = 0; i < expectedTokensWithoutBpt.length; i++) {
+            assertEq(address(tokens[i]), address(expectedTokensWithoutBpt[i]));
+        }
+    }
+
+    function testDropBptFromBalances(uint256[] memory balances) external {
```
Comment (nventuro) [inline pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol:35]: Isn't this the exact same?

Comment (TomAFrench) [reply pkg/pool-weighted/test/foundry/DropBptFromArray.t.sol:36]: Yeah, this is just for completeness.

**Pull request merged**